src/code/z_skin_awb.c

```diff
@@ -1,13 +1,204 @@
+            }
+        }
+    }
+}
+
+/**
+ * Initialises a skin skeleton to looping animation, dynamically allocating the frame tables,
+ * and dynamically allocating and initializing the Vtx and SkinLimbVtx buffers for its animated limbs
+ */
+void Skin_Init(GameState* gameState, Skin* skin, SkeletonHeader* skeletonHeader, AnimationHeader* animationHeader) {
+    s32 limbCount;
+    s32 i;
+    SkinLimb** skeleton;
+    SkeletonHeader* virtSkelHeader = Lib_SegmentedToVirtual(skeletonHeader);
+
+    skin->limbCount = virtSkelHeader->limbCount;
+    skin->skeletonHeader = virtSkelHeader;
+    limbCount = skin->skeletonHeader->limbCount;
+
+    skeleton = Lib_SegmentedToVirtual(skin->skeletonHeader->segment);
+    skin->vtxTable = ZeldaArena_Malloc(limbCount * sizeof(SkinLimbVtx));
+
+    for (i = 0; i < limbCount; i++) {
+        SkinLimbVtx* vtxEntry = &skin->vtxTable[i];
+
+        if ((((SkinLimb*)Lib_SegmentedToVirtual(skeleton[i]))->segmentType != SKIN_LIMB_TYPE_ANIMATED) ||
+            (((SkinLimb*)Lib_SegmentedToVirtual(skeleton[i]))->segment == NULL)) {
+            vtxEntry->index = 0;
+
+            vtxEntry->buf[0] = NULL;
+            vtxEntry->buf[1] = NULL;
+        } else {
+            SkinAnimatedLimbData* animatedLimbData =
+                Lib_SegmentedToVirtual((((SkinLimb*)Lib_SegmentedToVirtual(skeleton[i]))->segment));
+
+            { s32 tmp; }
+
+            vtxEntry->index = 0;
+            vtxEntry->buf[0] = ZeldaArena_Malloc(animatedLimbData->totalVtxCount * sizeof(Vtx));
+            vtxEntry->buf[1] = ZeldaArena_Malloc(animatedLimbData->totalVtxCount * sizeof(Vtx));
+
+            Skin_InitAnimatedLimb(gameState, skin, i);
+        }
+    }
+
+    SkelAnime_InitSkin(gameState, &skin->skelAnime, skeletonHeader, animationHeader);
+}
+
+/**
+ * Frees the dynamically allocated Vtx and SkinLimbVtx buffers and tables
+ */
+void Skin_Free(GameState* gameState, Skin* skin) {
+    if (skin->vtxTable != NULL) {
+        s32 i;
+
+        for (i = 0; i < skin->limbCount; i++) {
+            if (skin->vtxTable[i].buf[0] != NULL) {
+                ZeldaArena_Free(skin->vtxTable[i].buf[0]);
+                skin->vtxTable[i].buf[0] = NULL;
+            }
+            if (skin->vtxTable[i].buf[1] != NULL) {
+                ZeldaArena_Free(skin->vtxTable[i].buf[1]);
+                skin->vtxTable[i].buf[1] = NULL;
+            }
+        }
+
+        if (skin->vtxTable != NULL) {
+            ZeldaArena_Free(skin->vtxTable);
+        }
+
+        SkelAnime_Free(&skin->skelAnime, (GlobalContext*)gameState);
+    }
+}
+
+s32 func_801387D4(Skin* skin, SkinLimb** skeleton, MtxF* limbMatrices, u8 parentIndex, u8 limbIndex) {
+    s32 pad;
+    SkinLimb* limb = Lib_SegmentedToVirtual(skeleton[limbIndex]);
+    MtxF* mtx;
+    s32 ret;
+    MtxF sp28;
+
+    if (parentIndex == LIMB_DONE) {
+        SkinMatrix_GetClear(&mtx);
+    } else {
+        mtx = &limbMatrices[(s32)parentIndex];
```

Comment:
Cast required? Or argument is definitely a u8?

Reply:
The cast is required